File: core/src/main/java/javabot/javadoc/JavadocMethod.java
package javabot.javadoc;

import org.bson.types.ObjectId;
import org.mongodb.morphia.annotations.Entity;
import org.mongodb.morphia.annotations.Id;
import org.mongodb.morphia.annotations.Index;
import org.mongodb.morphia.annotations.Indexes;
import org.mongodb.morphia.annotations.PrePersist;

import java.util.List;

@Entity(value = "methods", noClassnameStored = true)
@Indexes({
    @Index("apiId"),
    @Index("javadocClassId, upperName"),
    @Index("apiId, javadocClassId, upperName"),
})
public class JavadocMethod extends JavadocElement {
  @Id
  private ObjectId id;
  private ObjectId javadocClassId;

  private String name;
  private String upperName;
  private String longSignatureTypes;
  private String shortSignatureTypes;
  private Integer paramCount;

  private String parentClassName;

  public JavadocMethod() {
  }

  public JavadocMethod(final JavadocClass parent, final String name, final int count,
      final List<String> longArgs, final List<String> shortArgs) {
    this.name = name;
    javadocClassId = parent.getId();
    setApiId(parent.getApiId());
    parentClassName = parent.toString();

    paramCount = count;
    longSignatureTypes = String.join(", ", longArgs);
    shortSignatureTypes = String.join(", ", shortArgs);
    buildUrl(parent, longArgs);
  }

  private void buildUrl(final JavadocClass parent, final List<String> longArgs) {
    String parentUrl = parent.getDirectUrl();
    boolean java8 = parentUrl.contains("se/8");
    StringBuilder url = new StringBuilder();
    for (String arg : longArgs) {
      if(url.length() != 0) {
        url.append(java8 ? "-": ", ");
      }
      url.append(arg.replaceAll("<.*", ""));
    }
    String directUrl = java8 ? parentUrl + "#" + this.name + "-" + url + "-" :
        parentUrl + "#" + this.name + "(" + url + ")";

    setLongUrl(directUrl);
    setDirectUrl(directUrl);
  }

  public ObjectId getId() {
    return id;
  }

  public ObjectId getJavadocClassId() {
    return javadocClassId;
  }

  public void setJavadocClassId(final ObjectId javadocClassId) {
    this.javadocClassId = javadocClassId;
  }

  public void setJavadocClassId(final JavadocClass javadocClass) {
    this.javadocClassId = javadocClass.getId();
    parentClassName = javadocClass.toString();
  }

  public void setId(final ObjectId methodId) {
    id = methodId;
  }

  public final String getShortSignature() {
    return name + "(" + shortSignatureTypes + ")";
  }

  public String getName() {
    return name;
  }

  public void setName(final String name) {
    this.name = name;
  }

  public String getLongSignatureTypes() {
    return longSignatureTypes;
  }

  public void setLongSignatureTypes(final String types) {
    longSignatureTypes = types;
  }

  public String getShortSignatureTypes() {
    return shortSignatureTypes;
  }

  public void setShortSignatureTypes(final String types) {
    shortSignatureTypes = types;
  }

  public Integer getParamCount() {
    return paramCount;
  }

  public void setParamCount(final Integer count) {
    paramCount = count;
  }

  public String getUpperName() {
    return upperName;
  }

  public void setUpperName(final String upperName) {
    this.upperName = upperName;
  }

  @PrePersist
  public void uppers() {
    upperName = name.toUpperCase();
  }

  @Override
  public String toString() {
    return parentClassName + "." + getShortSignature();
  }
}

File: src/main/kotlin/javabot/javadoc/JavadocMethod.kt
package javabot.javadoc

import org.bson.types.ObjectId
import org.mongodb.morphia.annotations.Entity
import org.mongodb.morphia.annotations.Field
import org.mongodb.morphia.annotations.Id
import org.mongodb.morphia.annotations.Index
import org.mongodb.morphia.annotations.Indexes
import org.mongodb.morphia.annotations.PrePersist
import org.mongodb.morphia.annotations.Reference

@Entity(value = "methods", noClassnameStored = true)
@Indexes(Index(fields = arrayOf(Field("apiId"))), Index(fields = arrayOf(Field("javadocClass"), Field("upperName") )),
      Index(fields = arrayOf(Field("apiId"), Field("javadocClass"), Field("upperName") )))
class JavadocMethod : JavadocElement {
    @Reference(lazy = true, idOnly = true)
    lateinit var javadocClass: JavadocClass
    lateinit var name: String
    lateinit var upperName: String
    lateinit var longSignatureTypes: String
    lateinit var shortSignatureTypes: String
    var paramCount: Int = 0

    lateinit var parentClassName: String

    constructor() {
    }

    constructor(parent: JavadocClass, name: String, count: Int,
                       longArgs: List<String>, shortArgs: List<String>) {
        this.name = name
        javadocClass = parent
        apiId = parent.apiId
        parentClassName = parent.toString()

        paramCount = count
        longArgs.joinToString(", ")
        longSignatureTypes = longArgs.joinToString(", ")
        shortSignatureTypes = shortArgs.joinToString(", ")
        buildUrl(parent, longArgs)
    }

    private fun buildUrl(parent: JavadocClass, longArgs: List<String>) {
        val parentUrl = parent.url
        val modern = parentUrl.contains("se/8") || parentUrl.contains("ee/7")
        val url = StringBuilder()
        for (arg in longArgs) {
            if (url.length != 0) {
                url.append(if (modern) "-" else ", ")
            }
            url.append(arg.replace("<.*?>".toRegex(), ""))
        }
        var directUrl = if (modern)
            parentUrl + "#" + this.name + "-" + url + "-"
        else
            parentUrl + "#" + this.name + "(" + url + ")"

        this.url = directUrl
    }

    fun getShortSignature(): String {
        return "$name($shortSignatureTypes)"
    }

    @PrePersist
    fun uppers() {
        upperName = name.toUpperCase()
    }

    override fun toString(): String {
        return parentClassName + "." + getShortSignature()
    }
}
